File: examples/log-analyzer.pl
#!/usr/bin/env perl
use strict;
use warnings;
binmode(STDOUT, ':encoding(UTF-8)');

# OpenClaw log analyzer (Perl) — parses gateway access logs from stdin
my %counts = (INFO => 0, WARN => 0, ERROR => 0);

while (my $line = <STDIN>) {
    if ($line =~ /^(\S+)\s+(INFO|WARN|ERROR)\s+(\S+)\s+(.+)$/) {
        my ($ts, $level, $node, $msg) = ($1, $2, $3, $4);
        $counts{$level}++;
        print "\x{26a0} $ts [$node] $msg\n" if $level eq 'ERROR';
    }
}

print "\n--- Summary ---\n";
print "$_: $counts{$_}\n" for qw(ERROR INFO WARN);

File: examples/log-analyzer.hs
-- OpenClaw log analyzer (Haskell) — parses gateway access logs from stdin
module Main where

import qualified Data.Map.Strict as Map

tally :: String -> Map.Map String Int -> Map.Map String Int
tally line acc =
  case words line of
    (_ : level : _) | level `elem` ["INFO", "WARN", "ERROR"] ->
      Map.insertWith (+) level 1 acc
    _ -> acc

main :: IO ()
main = do
  contents <- getContents
  let ls = lines contents
  mapM_ putStrLn
    [ "\9888 " ++ ts ++ " [" ++ node ++ "] " ++ unwords msg
    | l <- ls
    , (ts : "ERROR" : node : msg) <- [words l]
    ]
  putStrLn "\n--- Summary ---"
  mapM_ (\(k, v) -> putStrLn (k ++ ": " ++ show v))
    (Map.toList (foldr tally Map.empty ls))
